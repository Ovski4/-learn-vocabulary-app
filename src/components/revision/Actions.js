import React from 'react';
import { connect } from 'react-redux';
import { Text, Picker, Button, View, StyleSheet } from 'react-native';
import {
    translationsShuffled,
    translationsUnshuffled,
    allTranslationsRevealed,
    translationsHidden
} from '../../actions/uiTranslations';
import { translationsFilteredByTag } from '../../actions/ui';
import Header from '../../components/ui/Header';
import translationsService from '../../services/translations';

const getRandomNumbers = (arrayLength) => {
    const numbers = [];
    for (let i = 0; i < arrayLength; i++) {
        numbers.push(Math.random());
    }

    return numbers;
};

const mapStateToProps = (state) => ({
    translations: state.translations,
    tags: state.tags
});

const styles = StyleSheet.create({
    button: {
        flex: 1,
        margin: 5
    },
    revisionWords: {
        flexDirection: 'row',
        flex: 0.8
    },
    row: {
        flexDirection: 'row'
    },
    tagsSelectorRow: {
        flexDirection: 'row',
        alignItems: 'center',
        backgroundColor: '#fff',
        justifyContent: 'center'
    },
    picker:{
        height: 40,
        width: 200,
        paddingLeft: 20
    },
    filterText: {
        fontSize: 16,
        color: '#929292'
    }
});

const mapDispatchToProps = (dispatch) => ({
    handleShuffle: (translationsLength) => dispatch(translationsShuffled(
        getRandomNumbers(translationsLength)
    )),
    handleUnshuffle: () => dispatch(translationsUnshuffled()),
    handleDisplayEverything: (translations) => dispatch(allTranslationsRevealed(translations)),
    handleHideLeft: (translations) => dispatch(translationsHidden(translations, 'left')),
    handleHideRight: (translations) => dispatch(translationsHidden(translations, 'right')),
    handleFilterByTag: (tagId) => dispatch(translationsFilteredByTag(tagId))
});

class Actions extends React.PureComponent {

    constructor(props) {
        super(props);
        this.state = {
            selectedTag: null
        };
    }

    filterByTag = (selectedTag) => {
        this.setState({selectedTag});
        this.props.handleFilterByTag(selectedTag);
    }

    render() {

        const visibleTranslations = this.state.selectedTag !== null ?
            translationsService.filterByTagId(this.props.translations, this.state.selectedTag) :
            this.props.translations
        ;

        const pickerItems = this.props.tags
            .sort((tag1, tag2) => tag1.label < tag2.label ? -1 : 1)
            .map(tag => {
                return <Picker.Item key={tag.id} value={tag.id} label={tag.label} />
            })
        ;
        pickerItems.unshift(<Picker.Item key={'none'} value={null} label="No tag selected" />)

        return (
            <View>
                <Header>Actions</Header>
                <View style={styles.row}>
                    <View style={styles.button}>
                        <Button
                            onPress={() => this.props.handleShuffle(this.props.translations.length)}
                            title="Shuffle"
                        />
                    </View>
                    <View style={styles.button}>
                        <Button
                            disabled={translationsService.translationsAreOrdered(this.props.translations)}
                            onPress={() => this.props.handleUnshuffle()}
                            title="Unshuffle"
                        />
                    </View>
                </View>
                <View style={styles.row}>
                    <View style={styles.button}>
                        <Button
                            disabled={translationsService.translationsAreHidden(visibleTranslations, 'left')}
                            onPress={() => this.props.handleHideLeft(visibleTranslations)}
                            title="Hide left"
                        />
                    </View>
                    <View style={styles.button}>
                        <Button
                            disabled={translationsService.translationsAreHidden(visibleTranslations, 'right')}
                            onPress={() => this.props.handleHideRight(visibleTranslations)}
                            title="Hide right"
                        />
                    </View>
                    <View style={styles.button}>
                        <Button
                            disabled={translationsService.allTranslationsAreVisible(visibleTranslations)}
                            onPress={() => this.props.handleDisplayEverything(visibleTranslations)}
                            title="Display all"
                        />
                    </View>
                </View>
                <View style={styles.tagsSelectorRow}>
                    <Text style={styles.filterText}>Filter by:</Text>
                    <Picker
                        selectedValue={this.state.selectedTag}
                        style={styles.picker}
                        onValueChange={(tag) => {
                            // Reveal everything when we change the tag
                            this.props.handleDisplayEverything(visibleTranslations)
                            this.filterByTag(tag)
                        }}
                    >
                        {pickerItems}    
                    </Picker>
                </View>
            </View>
        );
    }
}

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(Actions);
