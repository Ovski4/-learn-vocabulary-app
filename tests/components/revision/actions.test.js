import reducer from '../../../src/reducers/index';
import { createStore } from 'redux';
import React from 'react';
import Actions from '../../../src/components/revision/actions/connect';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux'

describe('Actions component', () => {

    it('Should render the component', () => {

        const store = createStore(reducer, {
            config: {
                locale: 'en'
            }
        });

        const rendered = renderer.create(<Provider store={store}><Actions /></Provider>).toJSON();
        expect(rendered).toMatchSnapshot();
    });

});
