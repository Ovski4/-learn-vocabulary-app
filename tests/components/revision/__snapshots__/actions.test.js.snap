// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Actions component Should render the component with the english locale 1`] = `
<View>
  <View
    style={
      Object {
        "alignItems": "center",
        "margin": 5,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontWeight": "bold",
        }
      }
    >
      Actions
    </Text>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "margin": 5,
        }
      }
    >
      <View
        accessibilityComponentType="button"
        accessibilityTraits={
          Array [
            "button",
          ]
        }
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "padding": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            Shuffle
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "margin": 5,
        }
      }
    >
      <View
        accessibilityComponentType="button"
        accessibilityTraits={
          Array [
            "button",
            "disabled",
          ]
        }
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
              Object {},
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            disabled={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "padding": 8,
                  "textAlign": "center",
                },
                Object {
                  "color": "#cdcdcd",
                },
              ]
            }
          >
            Unshuffle
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "margin": 5,
        }
      }
    >
      <View
        accessibilityComponentType="button"
        accessibilityTraits={
          Array [
            "button",
            "disabled",
          ]
        }
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
              Object {},
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            disabled={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "padding": 8,
                  "textAlign": "center",
                },
                Object {
                  "color": "#cdcdcd",
                },
              ]
            }
          >
            Hide left
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "margin": 5,
        }
      }
    >
      <View
        accessibilityComponentType="button"
        accessibilityTraits={
          Array [
            "button",
            "disabled",
          ]
        }
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
              Object {},
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            disabled={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "padding": 8,
                  "textAlign": "center",
                },
                Object {
                  "color": "#cdcdcd",
                },
              ]
            }
          >
            Hide right
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "margin": 5,
        }
      }
    >
      <View
        accessibilityComponentType="button"
        accessibilityTraits={
          Array [
            "button",
            "disabled",
          ]
        }
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
              Object {},
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            disabled={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "padding": 8,
                  "textAlign": "center",
                },
                Object {
                  "color": "#cdcdcd",
                },
              ]
            }
          >
            Display all
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#929292",
          "fontSize": 16,
        }
      }
    >
      Filter by
      :
    </Text>
    <View
      style={
        Object {
          "height": 40,
          "paddingLeft": 20,
          "width": 200,
        }
      }
    >
      <RCTPicker
        items={
          Array [
            Object {
              "label": "No tag selected",
              "textColor": undefined,
              "value": null,
            },
          ]
        }
        onChange={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        selectedIndex={0}
        style={
          Array [
            Object {
              "height": 216,
            },
            undefined,
          ]
        }
      />
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#393e42",
          "borderBottomColor": "#000",
          "borderBottomWidth": 1,
          "borderTopColor": "#000",
          "borderTopWidth": 1,
        },
        Object {
          "backgroundColor": "#e1e8ee",
          "borderBottomColor": "#e1e1e1",
          "borderTopColor": "#e1e1e1",
        },
        Object {
          "backgroundColor": "transparent",
          "borderBottomColor": "transparent",
          "borderTopColor": "transparent",
        },
      ]
    }
  >
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      onClearText={[Function]}
      placeholder="Look for translations"
      placeholderTextColor="#86939e"
      style={
        Array [
          Object {
            "backgroundColor": "#303337",
            "borderRadius": 3,
            "color": "#86939e",
            "fontSize": 17.5,
            "height": 30,
            "margin": 8,
            "overflow": "hidden",
            "paddingLeft": 26,
            "paddingRight": 19,
          },
          Object {
            "backgroundColor": "#bdc6cf",
          },
          false,
          false,
          Object {
            "backgroundColor": "white",
            "borderColor": "#ddd",
            "borderWidth": 1,
            "marginLeft": 5,
            "marginRight": 5,
            "paddingLeft": 30,
          },
          false,
          Object {
            "paddingRight": 30,
          },
        ]
      }
      underlineColorAndroid="transparent"
    />
    <Text
      accessible={true}
      allowFontScaling={false}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "#86939e",
            "fontSize": 16,
          },
          Array [
            Object {
              "backgroundColor": "transparent",
              "position": "absolute",
              "top": 15.5,
            },
            Object {
              "left": 16,
            },
            undefined,
          ],
          Object {
            "fontFamily": "Material Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
        ]
      }
    >
      
    </Text>
    <Text
      accessible={true}
      allowFontScaling={false}
      ellipsizeMode="tail"
      onPress={[Function]}
      style={
        Array [
          Object {
            "color": "#86939e",
            "fontSize": 16,
          },
          Array [
            Object {
              "backgroundColor": "transparent",
              "position": "absolute",
              "top": 15.5,
            },
            Object {
              "right": 16,
            },
            undefined,
          ],
          Object {
            "fontFamily": "Material Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
        ]
      }
    >
      
    </Text>
  </View>
</View>
`;
